File: shared/whoop-client.ts
import type { TokenStore } from "./token-store.js";
import type { WhoopSleep, WhoopTokens, WhoopWorkout } from "./types.js";

const API_BASE = "https://api.prod.whoop.com/developer/v2";
const TOKEN_URL = "https://api.prod.whoop.com/oauth/oauth2/token";

export interface WhoopCredentials {
  clientId: string;
  clientSecret: string;
}

async function refreshWhoopTokens(creds: WhoopCredentials, refreshToken: string): Promise<WhoopTokens> {
  const res = await fetch(TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: creds.clientId,
      client_secret: creds.clientSecret,
      grant_type: "refresh_token",
      refresh_token: refreshToken,
    }),
  });
  if (!res.ok) {
    throw new Error(`WHOOP token refresh failed: ${res.status} ${await res.text()}`);
  }
  const body = (await res.json()) as { access_token: string; refresh_token: string; expires_in: number };
  return {
    access_token: body.access_token,
    refresh_token: body.refresh_token,
    expires_at: Date.now() + body.expires_in * 1000,
  };
}

/**
 * Returns a valid access token, refreshing and persisting to the shared KV
 * store if expired. WHOOP refresh tokens rotate on every use — this always
 * writes the new pair back to the store immediately after refreshing. The
 * same KV namespace is also read/written by the Python reconcile/rollup jobs,
 * so this is the single source of truth that avoids a refresh race between them.
 */
export async function getValidWhoopAccessToken(store: TokenStore, creds: WhoopCredentials): Promise<string> {
  const tokens = await store.getWhoopTokens();
  if (!tokens) {
    throw new Error("No WHOOP tokens in store — run `python -m scripts.whoop_auth` first.");
  }
  const EXPIRY_SKEW_MS = 60_000;
  if (tokens.expires_at - EXPIRY_SKEW_MS > Date.now()) {
    return tokens.access_token;
  }
  const refreshed = await refreshWhoopTokens(creds, tokens.refresh_token);
  await store.setWhoopTokens(refreshed);
  return refreshed.access_token;
}

export async function verifyWebhookSignature(
  rawBody: string,
  timestamp: string,
  signature: string,
  clientSecret: string,
): Promise<boolean> {
  const key = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(clientSecret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const mac = await crypto.subtle.sign("HMAC", key, new TextEncoder().encode(timestamp + rawBody));
  const expected = btoa(String.fromCharCode(...new Uint8Array(mac)));
  return expected === signature;
}

async function getJson<T>(path: string, accessToken: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  if (!res.ok) {
    throw new Error(`WHOOP API request failed (${path}): ${res.status} ${await res.text()}`);
  }
  return res.json() as Promise<T>;
}

export function getSleep(id: number, accessToken: string): Promise<WhoopSleep> {
  return getJson(`/activity/sleep/${id}`, accessToken);
}

export function getWorkout(id: number, accessToken: string): Promise<WhoopWorkout> {
  return getJson(`/activity/workout/${id}`, accessToken);
}

File: whoop_sync/whoop_client.py
import time
from typing import Optional

import requests

API_BASE = "https://api.prod.whoop.com/developer/v2"
TOKEN_URL = "https://api.prod.whoop.com/oauth/oauth2/token"
EXPIRY_SKEW_MS = 60_000


def exchange_code(client_id: str, client_secret: str, code: str, redirect_uri: str) -> dict:
    return _request_tokens(
        client_id,
        client_secret,
        {"grant_type": "authorization_code", "code": code, "redirect_uri": redirect_uri},
    )


def _refresh_tokens(client_id: str, client_secret: str, refresh_token: str) -> dict:
    return _request_tokens(
        client_id,
        client_secret,
        {"grant_type": "refresh_token", "refresh_token": refresh_token},
    )


def _request_tokens(client_id: str, client_secret: str, params: dict) -> dict:
    res = requests.post(
        TOKEN_URL,
        data={"client_id": client_id, "client_secret": client_secret, **params},
    )
    res.raise_for_status()
    body = res.json()
    return {
        "access_token": body["access_token"],
        "refresh_token": body["refresh_token"],
        "expires_at": time.time() * 1000 + body["expires_in"] * 1000,
    }


def get_valid_access_token(store, client_id: str, client_secret: str) -> str:
    """Refreshes and persists to the shared KV store if expired (see token_store.py)."""
    tokens = store.get_whoop_tokens()
    if not tokens:
        raise RuntimeError("No WHOOP tokens in store — run `python -m scripts.whoop_auth` first.")
    if tokens["expires_at"] - EXPIRY_SKEW_MS > time.time() * 1000:
        return tokens["access_token"]
    refreshed = _refresh_tokens(client_id, client_secret, tokens["refresh_token"])
    store.set_whoop_tokens(refreshed)
    return refreshed["access_token"]


def _get_json(path: str, access_token: str, params: Optional[dict] = None) -> dict:
    res = requests.get(
        f"{API_BASE}{path}",
        headers={"Authorization": f"Bearer {access_token}"},
        params=params,
    )
    res.raise_for_status()
    return res.json()


def get_sleep(record_id: int, access_token: str) -> dict:
    return _get_json(f"/activity/sleep/{record_id}", access_token)


def get_workout(record_id: int, access_token: str) -> dict:
    return _get_json(f"/activity/workout/{record_id}", access_token)


def _list_collection(path: str, access_token: str, start: str, end: str) -> list:
    records = []
    next_token = None
    while True:
        params = {"start": start, "end": end, "limit": 25}
        if next_token:
            params["nextToken"] = next_token
        page = _get_json(path, access_token, params)
        records.extend(page["records"])
        next_token = page.get("next_token")
        if not next_token:
            break
    return records


def list_sleep(access_token: str, start: str, end: str) -> list:
    return _list_collection("/activity/sleep", access_token, start, end)


def list_workouts(access_token: str, start: str, end: str) -> list:
    return _list_collection("/activity/workout", access_token, start, end)


def list_recovery(access_token: str, start: str, end: str) -> list:
    return _list_collection("/recovery", access_token, start, end)
